File: lab3/ProducerConsumerJava/src/com/company/Producer.java
package com.company;

public class Producer implements Runnable {
    private final int itemsToProducee;
    private final Manager manager;
    private final int id;

    public Producer(int itemsToproduce, int id, Manager manager) {
        this.itemsToProducee = itemsToproduce;
        this.id = id;
        this.manager = manager;
        new Thread(this, "Producer-" + id).start();
    }

    @Override
    public void run() {
        int produced = 0;
        while (produced < itemsToProducee) {
            if (!manager.registerProduced()) break;
            try {
                manager.full.acquire();
                manager.access.acquire();

                String item = "P" + id + "-item" + produced;
                manager.storage.add(item);
                System.out.printf("[Producer-%d] Added: %s  | storage: %d%n",
                        id, item, manager.storage.size());

                manager.access.release();
                manager.empty.release();
                produced++;

                Thread.sleep(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        System.out.printf("[Producer-%d] Done.%n", id);
    }
}

File: lab3/ProducerConsumer/ProducerConsumer/Producer.cs
using System;
using System.Threading;

namespace ProducerConsumer
{
    public class Producer
    {
        private readonly int itemsToProduce;
        private readonly int id;
        private readonly Manager manager;

        public Producer(int itemsToProduce, int id, Manager manager)
        {
            this.itemsToProduce = itemsToProduce;
            this.id = id;
            this.manager = manager;

            Thread thread = new Thread(Run);
            thread.Name = "Producer-" + id;
            thread.Start();
        }

        private void Run()
        {
            int produced = 0;
            while (produced < itemsToProduce)
            {
                if (!manager.RegisterProduced()) break; // глобальний ліміт вичерпано

                manager.Full.WaitOne();   // чекаю
                manager.Access.WaitOne(); // взаємне виключення

                string item = $"P{id}-item{produced}";
                manager.Storage.Add(item);
                Console.WriteLine($"[Producer-{id}] Added: {item}  | storage: {manager.Storage.Count}");

                manager.Access.Release();
                manager.Empty.Release();  // сигнал споживачу

                produced++;
                Thread.Sleep(500);
            }
            Console.WriteLine($"[Producer-{id}] Done.");
        }
    }
}